src/consensus/blockstore/slot_block_data.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright (c) Anza Technology, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Data structure holding shreds, slices and blocks for a specific slot.
+//!
+//!
+
+use std::collections::BTreeMap;
+
+use log::{debug, trace, warn};
+use thiserror::Error;
+
+use crate::consensus::votor::VotorEvent;
+use crate::crypto::signature::PublicKey;
+use crate::crypto::{Hash, MerkleTree};
+use crate::shredder::{self, RegularShredder, Shred, Shredder, Slice};
+use crate::{Block, Slot};
+
+use super::BlockInfo;
+
+///
+#[derive(Clone, Copy, Debug, Error)]
+pub enum AddShredError {
+    #[error("Shred has invalid signature")]
+    InvalidSignature,
+    #[error("Shred is already stored")]
+    Duplicate,
+    #[error("Shred shows leader equivocation")]
+    Equivocation,
+    #[error("Shred is after slice marked as last")]
+    AfterLastSlice,
+}
+
+///
+pub struct SlotBlockData {
+    /// Slot number this data corresponds to.
+    slot: Slot,
+    ///
+    pub(super) canonical: BlockData,
+    ///
+    pub(super) alternatives: BTreeMap<Hash, BlockData>,
+    /// Whether conflicting shreds have been seen for this slot.
+    pub(super) equivocated: bool,
+}
+
+pub struct BlockData {
+    ///
+    slot: Slot,
+    ///
+    pub(super) completed: Option<(Hash, Block)>,
+    ///
+    pub(super) shreds: BTreeMap<usize, Vec<Shred>>,
+    ///
+    pub(super) slices: BTreeMap<usize, Slice>,
+    ///
+    pub(super) last_slice: Option<usize>,
+    ///
+    pub(super) double_merkle_tree: Option<MerkleTree>,
+    ///
+    pub(super) merkle_root_cache: BTreeMap<usize, Hash>,
+}
+
+impl BlockData {
+    pub fn new(slot: Slot) -> Self {
+        Self {
+            slot,
+            completed: None,
+            shreds: BTreeMap::new(),
+            slices: BTreeMap::new(),
+            last_slice: None,
+            double_merkle_tree: None,
+            merkle_root_cache: BTreeMap::new(),
+        }
+    }
+}
+
+impl SlotBlockData {
+    pub fn new(slot: Slot) -> Self {
+        Self {
+            slot,
+            canonical: BlockData::new(slot),
+            alternatives: BTreeMap::new(),
+            equivocated: false,
+        }
+    }
+
+    ///
+    pub fn add_shred_from_disseminator(
+        &mut self,
+        shred: Shred,
+        leader_pk: PublicKey,
+    ) -> Result<Option<VotorEvent>, AddShredError> {
+        assert_eq!(shred.payload().slot, self.slot);
+        if self.equivocated {
+            debug!("recevied shred from equivocating leader, not adding to blockstore");
+            return Err(AddShredError::Equivocation);
+        }
+        let add_shred_result = self.canonical.check_shred_to_add(&shred, true, leader_pk);
+        if matches!(add_shred_result, Err(AddShredError::Equivocation)) {
+            self.equivocated = true;
+        }
+        add_shred_result?;
```

Comment:
If `check_shred_to_add` ever starts returning something other than `()` in the `Ok` case, the compiler will help us catch this place to update.  Maybe we can write the above code as something like this:

```
match check_shred_to_add(...) {
  Ok(()) => Ok(self.canonical.add_valid_shred(shred)),
  Err(err) => {
    if let AddShredError::Equivocation = &err {
       self.equivocated = true;
    }
    Err(err)
  }
}
```